// @mui/joy/node/styles/CssVarsProvider.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useColorScheme = exports.getInitColorSchemeScript = exports.CssVarsProvider = void 0;

var _utils = require("@mui/utils");

var _system = require("@mui/system");

var _extendTheme = _interopRequireDefault(require("./extendTheme"));

var _variantUtils = require("./variantUtils");

const {
  CssVarsProvider,
  useColorScheme,
  getInitColorSchemeScript
} = (0, _system.unstable_createCssVarsProvider)({
  theme: (0, _extendTheme.default)(),
  attribute: 'data-joy-color-scheme',
  modeStorageKey: 'joy-mode',
  colorSchemeStorageKey: 'joy-color-scheme',
  defaultColorScheme: {
    light: 'light',
    dark: 'dark'
  },
  prefix: 'joy',
  resolveTheme: mergedTheme => {
    mergedTheme.variants = (0, _utils.deepmerge)({
      plain: (0, _variantUtils.createVariant)('plain', mergedTheme),
      plainHover: (0, _variantUtils.createVariant)('plainHover', mergedTheme),
      plainActive: (0, _variantUtils.createVariant)('plainActive', mergedTheme),
      plainDisabled: (0, _variantUtils.createVariant)('plainDisabled', mergedTheme),
      outlined: (0, _variantUtils.createVariant)('outlined', mergedTheme),
      outlinedHover: (0, _variantUtils.createVariant)('outlinedHover', mergedTheme),
      outlinedActive: (0, _variantUtils.createVariant)('outlinedActive', mergedTheme),
      outlinedDisabled: (0, _variantUtils.createVariant)('outlinedDisabled', mergedTheme),
      soft: (0, _variantUtils.createVariant)('soft', mergedTheme),
      softHover: (0, _variantUtils.createVariant)('softHover', mergedTheme),
      softActive: (0, _variantUtils.createVariant)('softActive', mergedTheme),
      softDisabled: (0, _variantUtils.createVariant)('softDisabled', mergedTheme),
      solid: (0, _variantUtils.createVariant)('solid', mergedTheme),
      solidHover: (0, _variantUtils.createVariant)('solidHover', mergedTheme),
      solidActive: (0, _variantUtils.createVariant)('solidActive', mergedTheme),
      solidDisabled: (0, _variantUtils.createVariant)('solidDisabled', mergedTheme),
      // variant overrides
      plainOverrides: (0, _variantUtils.createTextOverrides)(mergedTheme),
      outlinedOverrides: (0, _variantUtils.createTextOverrides)(mergedTheme),
      softOverrides: (0, _variantUtils.createTextOverrides)(mergedTheme),
      solidOverrides: (0, _variantUtils.createContainedOverrides)(mergedTheme)
    }, mergedTheme.variants, {
      clone: false
    });
    return mergedTheme;
  },
  shouldSkipGeneratingVar: keys => keys[0] === 'typography' || keys[0] === 'variants' || keys[0] === 'focus' || keys[0] === 'breakpoints'
});
exports.getInitColorSchemeScript = getInitColorSchemeScript;
exports.useColorScheme = useColorScheme;
exports.CssVarsProvider = CssVarsProvider;
