// @mui/joy/node/styles/defaultTheme.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getThemeWithVars = exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _utils = require("@mui/utils");

var _extendTheme2 = _interopRequireDefault(require("./extendTheme"));

var _variantUtils = require("./variantUtils");

const _excluded = ["colorSchemes", "fontFamily", "fontSize", "fontWeight", "letterSpacing", "lineHeight", "radius", "shadow", "palette"],
      _excluded2 = ["mode", "colorScheme"];

const getThemeWithVars = themeInput => {
  const _extendTheme = (0, _extendTheme2.default)(themeInput),
        {
    colorSchemes,
    fontFamily,
    fontSize,
    fontWeight,
    letterSpacing,
    lineHeight,
    radius,
    shadow,
    palette: paletteInput
  } = _extendTheme,
        restTheme = (0, _objectWithoutPropertiesLoose2.default)(_extendTheme, _excluded);

  const colorSchemePalette = (0, _utils.deepmerge)(colorSchemes[(paletteInput == null ? void 0 : paletteInput.colorScheme) || 'light'].palette, paletteInput);
  const _ref = colorSchemePalette,
        {
    mode = 'light',
    colorScheme = 'light'
  } = _ref,
        palette = (0, _objectWithoutPropertiesLoose2.default)(_ref, _excluded2);
  const defaultTheme = (0, _extends2.default)({
    fontFamily,
    fontSize,
    fontWeight,
    letterSpacing,
    lineHeight,
    radius,
    shadow
  }, restTheme, {
    colorSchemes: (0, _extends2.default)({}, colorSchemes, {
      [colorScheme]: palette
    }),
    palette: (0, _extends2.default)({}, palette, {
      mode,
      colorScheme
    }),
    vars: {
      fontFamily,
      fontSize,
      fontWeight,
      letterSpacing,
      lineHeight,
      radius,
      shadow,
      palette
    }
  });
  defaultTheme.variants = (0, _utils.deepmerge)({
    plain: (0, _variantUtils.createVariant)('plain', defaultTheme),
    plainHover: (0, _variantUtils.createVariant)('plainHover', defaultTheme),
    plainActive: (0, _variantUtils.createVariant)('plainActive', defaultTheme),
    plainDisabled: (0, _variantUtils.createVariant)('plainDisabled', defaultTheme),
    outlined: (0, _variantUtils.createVariant)('outlined', defaultTheme),
    outlinedHover: (0, _variantUtils.createVariant)('outlinedHover', defaultTheme),
    outlinedActive: (0, _variantUtils.createVariant)('outlinedActive', defaultTheme),
    outlinedDisabled: (0, _variantUtils.createVariant)('outlinedDisabled', defaultTheme),
    soft: (0, _variantUtils.createVariant)('soft', defaultTheme),
    softHover: (0, _variantUtils.createVariant)('softHover', defaultTheme),
    softActive: (0, _variantUtils.createVariant)('softActive', defaultTheme),
    softDisabled: (0, _variantUtils.createVariant)('softDisabled', defaultTheme),
    solid: (0, _variantUtils.createVariant)('solid', defaultTheme),
    solidHover: (0, _variantUtils.createVariant)('solidHover', defaultTheme),
    solidActive: (0, _variantUtils.createVariant)('solidActive', defaultTheme),
    solidDisabled: (0, _variantUtils.createVariant)('solidDisabled', defaultTheme),
    // variant overrides
    plainOverrides: (0, _variantUtils.createTextOverrides)(defaultTheme),
    outlinedOverrides: (0, _variantUtils.createTextOverrides)(defaultTheme),
    softOverrides: (0, _variantUtils.createTextOverrides)(defaultTheme),
    solidOverrides: (0, _variantUtils.createContainedOverrides)(defaultTheme)
  }, defaultTheme.variants);
  return defaultTheme;
};

exports.getThemeWithVars = getThemeWithVars;
const defaultTheme = getThemeWithVars();
var _default = defaultTheme;
exports.default = _default;
