// @mui/joy/node/List/List.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _clsx = _interopRequireDefault(require("clsx"));

var _utils = require("@mui/utils");

var _composeClasses = _interopRequireDefault(require("@mui/base/composeClasses"));

var _styles = require("../styles");

var _listClasses = require("./listClasses");

var _NestedListContext = _interopRequireDefault(require("./NestedListContext"));

var _RowListContext = _interopRequireDefault(require("./RowListContext"));

var _ComponentListContext = _interopRequireDefault(require("./ComponentListContext"));

var _jsxRuntime = require("react/jsx-runtime");

const _excluded = ["component", "className", "children", "size", "row"];

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const useUtilityClasses = ownerState => {
  const {
    size,
    nested,
    row
  } = ownerState;
  const slots = {
    root: ['root', size && `size${(0, _utils.unstable_capitalize)(size)}`, nested && 'nested', row && 'row']
  };
  return (0, _composeClasses.default)(slots, _listClasses.getListUtilityClass, {});
};

const ListRoot = (0, _styles.styled)('ul', {
  name: 'JoyList',
  slot: 'Root',
  overridesResolver: (props, styles) => styles.root
})(({
  theme,
  ownerState
}) => {
  function applySizeVars(size) {
    if (size === 'sm') {
      return {
        '--List-gap': '0.25rem',
        '--List-item-minHeight': '2rem',
        '--List-item-paddingY': '0.25rem',
        '--List-item-paddingLeft': '0.25rem',
        '--List-item-paddingRight': '0.25rem',
        '--List-item-fontSize': theme.vars.fontSize.sm,
        '--List-decorator-width': ownerState.row ? '1.5rem' : '2rem',
        '--Icon-fontSize': '1.125rem'
      };
    }

    if (size === 'md') {
      return {
        '--List-gap': '0.375rem',
        '--List-item-minHeight': '2.5rem',
        '--List-item-paddingY': '0.375rem',
        '--List-item-paddingLeft': '0.75rem',
        '--List-item-paddingRight': '0.75rem',
        '--List-item-fontSize': theme.vars.fontSize.md,
        '--List-decorator-width': ownerState.row ? '2rem' : '2.5rem',
        '--Icon-fontSize': '1.25rem'
      };
    }

    if (size === 'lg') {
      return {
        '--List-gap': '0.5rem',
        '--List-item-minHeight': '3rem',
        '--List-item-paddingY': '0.5rem',
        '--List-item-paddingLeft': '0.5rem',
        '--List-item-paddingRight': '0.5rem',
        '--List-item-fontSize': theme.vars.fontSize.md,
        '--List-decorator-width': ownerState.row ? '2.5rem' : '3rem',
        '--Icon-fontSize': '1.5rem'
      };
    }

    return {};
  }

  return [ownerState.nested && (0, _extends2.default)({}, applySizeVars(ownerState.instanceSize), {
    '--List-item-paddingLeft': 'var(--NestedList-item-paddingLeft)',
    // reset ListItem, ListItemButton negative margin (caused by NestedListItem)
    '--List-itemButton-margin': '0px',
    '--List-item-margin': '0px',
    padding: 0,
    margin: 'var(--NestedList-margin)',
    marginTop: 'var(--List-gap)'
  }), !ownerState.nested && (0, _extends2.default)({}, applySizeVars(ownerState.size), {
    '--List-padding': '0px',
    '--List-radius': '0px',
    '--List-divider-gap': '0px',
    '--List-decorator-color': theme.vars.palette.text.tertiary,
    '--List-nestedInsetStart': '0px',
    // by default, The ListItem & ListItemButton use automatic radius adjustment based on the parent List.
    '--List-item-radius': 'max(var(--List-radius) - var(--List-padding), min(var(--List-padding) / 2, var(--List-radius) / 2))',
    '--List-item-startActionTranslateX': 'var(--List-item-paddingLeft)',
    '--List-item-endActionTranslateX': 'calc(-1 * var(--List-item-paddingLeft))',
    borderRadius: 'var(--List-radius)',
    padding: 'var(--List-padding)',
    margin: 'initial'
  }), {
    listStyle: 'none',
    display: 'flex',
    flexDirection: ownerState.row ? 'row' : 'column',
    flexGrow: 1,
    position: 'relative' // for sticky ListItem

  }];
});
const List = /*#__PURE__*/React.forwardRef(function List(inProps, ref) {
  const nested = React.useContext(_NestedListContext.default);
  const props = (0, _styles.useThemeProps)({
    props: inProps,
    name: 'JoyList'
  });
  const {
    component,
    className,
    children,
    size = 'md',
    row = false
  } = props,
        other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  const ownerState = (0, _extends2.default)({
    instanceSize: inProps.size,
    size,
    nested,
    row
  }, props);
  const classes = useUtilityClasses(ownerState);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_RowListContext.default.Provider, {
    value: row,
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_ComponentListContext.default.Provider, {
      value: typeof component === 'string' ? component : undefined,
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(ListRoot, (0, _extends2.default)({
        ref: ref,
        as: component,
        className: (0, _clsx.default)(classes.root, className),
        ownerState: ownerState
      }, other, {
        children: children
      }))
    })
  });
});
process.env.NODE_ENV !== "production" ? List.propTypes
/* remove-proptypes */
= {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // |     To update them edit TypeScript types and run "yarn proptypes"  |
  // ----------------------------------------------------------------------

  /**
   * The content of the component.
   */
  children: _propTypes.default.node,

  /**
   * Override or extend the styles applied to the component.
   */
  classes: _propTypes.default.object,

  /**
   * @ignore
   */
  className: _propTypes.default.string,

  /**
   * The component used for the root node.
   * Either a string to use a HTML element or a component.
   */
  component: _propTypes.default.elementType,

  /**
   * If `true`, display the list in horizontal direction.
   */
  row: _propTypes.default.bool,

  /**
   * The size of the component (affect other nested list* components).
   */
  size: _propTypes.default
  /* @typescript-to-proptypes-ignore */
  .oneOfType([_propTypes.default.oneOf(['sm', 'md', 'lg']), _propTypes.default.string]),

  /**
   * The system prop that allows defining system overrides as well as additional CSS styles.
   */
  sx: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.object, _propTypes.default.bool])), _propTypes.default.func, _propTypes.default.object])
} : void 0;
var _default = List;
exports.default = _default;
